ext/mbstring/mbstring.c

```diff
@@ -2941,6 +2941,130 @@ PHP_FUNCTION(mb_strtolower)
 	RETURN_STR(mbstring_convert_case(PHP_UNICODE_CASE_LOWER, ZSTR_VAL(str), ZSTR_LEN(str), enc));
 }
 
+static bool is_trim_wchar(uint32_t w, uint32_t *characters, size_t characters_len)
+{
+	for (size_t i = 0; i < characters_len; i++) {
+		if (w == characters[i]) {
+			return true;
+		}
+	}
+	return false;
+}
+
+static zend_string* trim_each_wchar(zend_string *str, zend_string *what, int mode, const mbfl_encoding *enc)
+{
+	unsigned char *in = (unsigned char*)ZSTR_VAL(str);
+	unsigned char *what_in = (unsigned char*)ZSTR_VAL(what);
+	uint32_t wchar_buf[128];
+	uint32_t what_wchar_buf[128];
+	size_t in_len = ZSTR_LEN(str);
+	size_t out_len = 0;
+	size_t what_out_len = 0;
+	size_t what_len = ZSTR_LEN(what);
+	unsigned int state = 0;
+	long left = 0;
+	long right = 0;
+	size_t total_len = 0;
+	bool first_left = true;
+
+	while (in_len) {
+		out_len = enc->to_wchar(&in, &in_len, wchar_buf, 128, &state);
+		ZEND_ASSERT(out_len <= 128);
+		total_len += out_len;
+
+		while (what_len) {
+			what_out_len = enc->to_wchar(&what_in, &what_len, what_wchar_buf, 128, &state);
+			ZEND_ASSERT(what_out_len <= 128);
+
+			if (first_left && (mode & 1)) {
+				for (size_t i = 0; i < out_len; i++) {
+					uint32_t w = wchar_buf[i];
+					if (is_trim_wchar(w, what_wchar_buf, what_out_len)) {
+						left += 1;
+					} else {
+						break;
+					}
+				}
+				first_left = false;
+			}
+
+			if (in_len == 0 && mode & 2) {
+				size_t i = out_len - 1;
+				while (i > 0) {
+					uint32_t w = wchar_buf[i];
+					i--;
+					if (is_trim_wchar(w, what_wchar_buf, what_out_len)) {
+						right -= 1;
+					} else {
+						break;
+					}
+				}
+			}
+		}
+	}
+	zend_string_release(str);
+	zend_string_release(what);
+
+	return mb_get_substr(str, left, total_len + (right - left), enc);
+}
+
+static zend_string* mb_trim_default_chars(zend_string *str, int mode, const mbfl_encoding *enc)
+{
+	uint32_t trim_default_chars[] = {
+		0x20, 0x0A, 0x0D, 0x09, 0x0B, 0x00, 0xA0, 0x1680,
+		0x2000, 0x2001, 0x2002, 0x2003, 0x2004, 0x2005, 0x2006, 0x2007,
+		0x2008, 0x2009, 0x200A, 0x2028, 0x2029, 0x202F, 0x205F, 0x3000,
+		0x85, 0x180E
+	};
+	size_t trim_default_chars_length = sizeof(trim_default_chars) / sizeof(uint32_t);
+	mb_convert_buf buf;
+	mb_convert_buf_init(&buf, trim_default_chars_length, MBSTRG(current_filter_illegal_substchar), MBSTRG(current_filter_illegal_mode));
+
+	enc->from_wchar(trim_default_chars, trim_default_chars_length, &buf, true);
+
+	return trim_each_wchar(str, mb_convert_buf_result(&buf, enc), mode, enc);
+}
+
+static zend_always_inline void php_do_mb_trim(INTERNAL_FUNCTION_PARAMETERS, int mode)
+{
+	zend_string *str;
+	zend_string *what = NULL;
+	zend_string *encoding = NULL;
+
+	ZEND_PARSE_PARAMETERS_START(1, 3)
+		Z_PARAM_STR(str)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_STR_OR_NULL(what)
+		Z_PARAM_STR_OR_NULL(encoding)
+	ZEND_PARSE_PARAMETERS_END();
+
+	const mbfl_encoding *enc = php_mb_get_encoding(encoding, 5);
```

Comment:
The argument number should be 3, not 5.

Reply:
Thank you very much! Fixed and add test.